src/auth/index.ts

```diff
@@ -62,23 +75,47 @@ export class Authenticator {
   }
 
   private setToken(token: string, expiresIn: number) {
+    // Set expiration to half of the expiresIn value
+    // to allow for some buffer time before the token expires
+    const tokenExpiryTimestampMs = Date.now() + (expiresIn * 1000) / 2;
     this.accessToken = token;
+    this.tokenExpiryTimestampMs = tokenExpiryTimestampMs;
+    // Update cache
     const key = this.getCacheKey();
-    key && this.getCache().set(key, { token, expiration: expiresIn });
+    key &&
+      this.getCache().set(key, {
+        token,
+        tokenExpiryTimestampMs
+      });
   }
 
-  private getCachedToken(): string | undefined {
+  private getCachedTokenRecord(): TokenRecord | undefined {
     const key = this.getCacheKey();
-    return key ? this.getCache().get(key)?.token : undefined;
-  }
-
-  getHeaders() {
-    if (this.accessToken) {
+    if (!key) return undefined;
+
+    const cachedTokenRecord = this.getCache().get(key);
+    // Check if token exists and is not expired
+    if (
+      cachedTokenRecord &&
+      Date.now() < cachedTokenRecord.tokenExpiryTimestampMs
+    ) {
       return {
-        Authorization: `Bearer ${this.accessToken}`
+        token: cachedTokenRecord.token,
+        tokenExpiryTimestampMs: cachedTokenRecord.tokenExpiryTimestampMs
       };
     }
-    return {};
+
+    return undefined;
+  }
+
+  async getToken(): Promise<string | undefined> {
+    if (
+      this.tokenExpiryTimestampMs &&
+      this.tokenExpiryTimestampMs < Date.now()
+    ) {
+      await this.authenticate();
```

Comment:
Should we also authenticate if `this.accessToken` is undefined?

Reply:
Technically it should never happen, but I can add a check to be on a safe side.